### PR TITLE
fix envoy process exit bug

### PR DESCRIPTION
**Pull request by silenceshell**
Running istio(pilot-agent and envoy) on virtual machine, when systemctl stop istio, the istio-pilot process is terminated, but the envoy process is still running.

(There is another fix about systemd #33040 )

`envoy.Agent` in istio-agent send `errAbort` to `a.abortCh`.

```go
func (a *Agent) terminate() {
	log.Infof("Agent draining Proxy")
	e := a.proxy.Drain()
	if e != nil {
		log.Warnf("Error in invoking drain listeners endpoint %v", e)
	}
	log.Infof("Graceful termination period is %v, starting...", a.terminationDrainDuration)
	time.Sleep(a.terminationDrainDuration)
	log.Infof("Graceful termination period complete, terminating remaining proxies.")
	a.abortCh <- errAbort         //<--- sending errAbort to abortCh
	log.Warnf("Aborted all epochs")
}
```

And expect envoy goroutine in istio-agent to kill envoy process when receiving err from abort.

```go
func (e *envoy) Run(epoch int, abort <-chan error) error {
..
	select {
	case err := <-abort:     //<---- receive abort err in envoy gorouting
		log.Warnf("Aborting epoch %d", epoch)
		if errKill := cmd.Process.Kill(); errKill != nil {
			log.Warnf("killing epoch %d caused an error %v", epoch, errKill)
		}
		return err
	case err := <-done:
		return err
	}
```

However, the main envoy.Agent returns immediately without waiting envoy goroutine. This may causes the envoy process is still running.

```go
// Run starts the envoy and waits until it terminates.
func (a *Agent) Run(ctx context.Context) {
	log.Info("Starting proxy agent")
	go a.runWait(0, a.abortCh)

	select {
	case status := <-a.statusCh:
..
	case <-ctx.Done():
		a.terminate()
		log.Info("Agent has successfully terminated")      //<---here returns immediately without waiting for a.runWait(0, a.abortCh)
	}
}
```

[X] Networking

Pull Request Attributes

Please check any characteristics that apply to this pull request. 

[X] Does not have any [user-facing](https://github.com/istio/istio/tree/master/releasenotes#when-to-add-release-notes) changes. This may include CLI changes, API changes, behavior changes, performance improvements, etc.

### CODE REVIEW

**Review by hzxuzhonghu — COMMENTED** [review]
I am not sure this makes sense as we terminate it initiatively

File: pkg/envoy/agent.go [hunk 1]
```diff
@@ -89,6 +89,12 @@ func (a *Agent) Run(ctx context.Context) {
 		log.Infof("No more active epochs, terminating")
 	case <-ctx.Done():
 		a.terminate()
+		status := <-a.statusCh
+		if status.err != nil && status.err == errAbort {
```
Comment (hzxuzhonghu) [inline pkg/envoy/agent.go:93]:
```suggestion
		if status.err == errAbort {
```

**Review by hzxuzhonghu — COMMENTED** [review]
Understood, it has to wait for envoy exit after calling terminate.

**Review by hzxuzhonghu — APPROVED** [review]

**Comment by silenceshell** [comment]
@hzxuzhonghu thanks for your quick response!

There are one problems here, even though we terminate it initiatively:
1. There is a certain probability when `systemctl stop istio` pilot-agent is killed but envoy process is still running, I have to kill envoy process with `kill` command manually.
2. As pilot-agent has terminated without waiting for the goroutine to send kill to envoy process, thus may cause the envoy still working with no manager.

**Comment by silenceshell** [comment]
@googlebot I fixed it.

**Comment by silenceshell** [comment]
@googlebot I fixed it.

**Comment by silenceshell** [comment]
@googlebot I fixed it.

**Comment by silenceshell** [comment]
@googlebot I fixed it.

**Comment by silenceshell** [comment]
FYI: If you met with the same problem, try to enable both this PR and #33040.

**Comment by S-Chan** [comment]
@hzxuzhonghu can we pick this back to 1.10?

**Comment by hzxuzhonghu** [comment]
I agree, we should backport all the envoy exit bug fixes. What other prs, have they been backportted?

**Comment by S-Chan** [comment]
https://github.com/istio/istio/pull/33040 seems to be the other one.

**Pull request merged**